lib/rubocop/cop/performance/literal_in_loop.rb

```diff
@@ -0,0 +1,140 @@
+# frozen_string_literal: true
+
+require 'set'
+
+module RuboCop
+  module Cop
+    module Performance
+      # This cop identifies places where Array and Hash literals are used
+      # within loops. It is better to extract them into a local variable or constant
+      # to avoid unnecessary allocations on each iteration.
+      #
+      # You can set the minimum number of elements to consider
+      # an offense with `MinSize`.
+      #
+      # @example
+      #   # bad
+      #   users.select do |user|
+      #     %i[superadmin admin].include?(user.role)
+      #   end
+      #
+      #   # good
+      #   admin_roles = %i[superadmin admin]
+      #   users.select do |user|
+      #     admin_roles.include?(user.role)
+      #   end
+      #
+      #   # good
+      #   ADMIN_ROLES = %i[superadmin admin]
+      #   ...
+      #   users.select do |user|
+      #     ADMIN_ROLES.include?(user.role)
+      #   end
+      #
+      class LiteralInLoop < Cop
+        MSG = 'Avoid immutable %<literal_class>s literals in loops. '\
+          'It is better to extract it into a local variable or a constant.'
+
+        POST_CONDITION_LOOP_TYPES = %i[while_post until_post].freeze
+        LOOP_TYPES = (POST_CONDITION_LOOP_TYPES + %i[while until for]).freeze
+
+        ENUMERABLE_METHOD_NAMES = (Enumerable.instance_methods + [:each]).to_set.freeze
```

Comment:
I have a feeling that the logic for immutable collections was needed somewhere else as well, so maybe it should eventually be made reusable in `rubocop-ast`.

Reply:
1. Yes, I have found something similar https://github.com/rubocop-hq/rubocop/blob/b571ba252fba0e13331871b065b28f27c107eeda/lib/rubocop/cop/lint/void.rb#L56-L62

2. And I already used this `ENUMERABLE_METHOD_NAMES` constant in another PR https://github.com/rubocop-hq/rubocop-performance/pull/127/files#diff-6e1da1392214a67f5b4f8919f93f9b20R31 , so worth extracting

So how do you feel about:
3. extracting those `POST_CONDITION_LOOP_TYPES` and `LOOP_TYPES` into some properly named constants in `rubocop-ast`s `AST::Node` 

4. and those `ENUMERABLE_METHOD_NAMES`, `NONMUTATING_ARRAY_METHODS` and `NONMUTATING_HASH_METHODS` into `rubocop-ast`s `AST::CollectionNode` module?

5. And probably it would be useful to extract those https://github.com/rubocop-hq/rubocop/blob/b571ba252fba0e13331871b065b28f27c107eeda/lib/rubocop/cop/lint/void.rb#L52-L54 into `AST::Node`.

6. And probably we should extend `AST::Node` with missing keywords to replace this hardcoded list? https://github.com/rubocop-hq/rubocop/blob/b571ba252fba0e13331871b065b28f27c107eeda/lib/rubocop/cop/style/redundant_self.rb#L146-L150

Reply:
3-6 all make sense to me.